src/bootman/config.c

```diff
@@ -58,44 +58,92 @@ bool boot_manager_set_timeout_value(BootManager *self, int timeout)
                 return false;
         }
 
-        if (fprintf(fp, "%d\n", timeout) < 0) {
+        if (fprintf(fp, "%s\n", contents) < 0) {
                 LOG_FATAL("Unable to set new timeout: %s", strerror(errno));
                 return false;
         }
         return true;
 }
 
-int boot_manager_get_timeout_value(BootManager *self)
+bool boot_manager_set_timeout_value(BootManager *self, int timeout)
+{
+        if (timeout <= 0) {
+            return _write_sysconf_file(self, "timeout", NULL);
+        }
+
+        autofree(char) *timeout_s = string_printf("%d", timeout);
+
+        return _write_sysconf_file(self, "timeout", timeout_s);
+}
+
+bool boot_manager_set_console_mode(BootManager *self, const char *mode)
+{
+        if (mode == NULL) {
+            return _write_sysconf_file(self, "console_mode", NULL);
+        }
+
+        return _write_sysconf_file(self, "console_mode", mode);
+}
+
+static bool _read_sysconf_value(BootManager *self, const char *filename, char *buf, size_t len)
 {
         autofree(FILE) *fp = NULL;
         autofree(char) *path = NULL;
-        int t_val;
+
+        char format[32];
+        snprintf(format, sizeof format, "%%%zu[^\n]", len);
 
         if (!self || !self->sysconfig) {
                 return false;
         }
 
-        path = string_printf("%s%s/timeout", self->sysconfig->prefix, KERNEL_CONF_DIRECTORY);
-
-        /* Default timeout being -1, i.e. don't use one */
+        path = string_printf("%s%s/%s", self->sysconfig->prefix, KERNEL_CONF_DIRECTORY, filename);
         if (!nc_file_exists(path)) {
-                return -1;
+                return false;
         }
 
         fp = fopen(path, "r");
         if (!fp) {
                 LOG_FATAL("Unable to open %s for reading: %s", path, strerror(errno));
-                return -1;
+                return false;
+        }
+
+        if (fscanf(fp, format, buf) != 1) {
+                LOG_ERROR("Failed to parse config file %s, using defaults", path);
+                return false;
+        }
+
+        return true;
+}
+
+int boot_manager_get_timeout_value(BootManager *self)
+{
+        char buf[32] = { 0 };
+        int t_val;
+
+        if (!_read_sysconf_value(self, "timeout", buf, sizeof buf)) {
+            return -1;
         }
 
-        if (fscanf(fp, "%d\n", &t_val) != 1) {
+        if (sscanf(buf, "%d\n", &t_val) != 1) {
                 LOG_ERROR("Failed to parse config file, defaulting to no timeout");
                 return -1;
         }
 
         return t_val;
 }
 
+char *boot_manager_get_console_mode(BootManager *self)
+{
+        char *value = (char*)malloc(64);
```

Comment:
I'm not sure I like this use of magic buffer lengths and mallocs. Also note this isn't a zero initialised (calloc) call, so any failures will result in a garbage value floating around, i.e. not `NULL`. Would it not be better to simply read a single line from a file (`getline` on an fp), free the internal getline pointer due to alloca crap in some implementations, then strdup the return. The console-mode check could just directly wrap it, whereas the timeout mode would simply pull it in as an `autofree()` and convert to a number (`atoi`)

Reply:
> I'm not sure I like this use of magic buffer lengths and mallocs.

I definitely don't. I changed it as suggested.